File: app/src/main/java/com/elvotra/clean/data/repository/PostsRepository.java
package com.elvotra.clean.data.repository;

import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;

import com.elvotra.clean.domain.model.Post;
import com.elvotra.clean.domain.repository.IPostsRepository;

import java.util.List;

public class PostsRepository implements IPostsRepository {

    private IPostsRepository postsRemoteRepository;
    private IPostsRepository postsLocalRepository;

    private static PostsRepository INSTANCE;

    public static PostsRepository getInstance(IPostsRepository postsRemoteRepository, IPostsRepository postsLocalRepository) {
        if (INSTANCE == null) {
            INSTANCE = new PostsRepository(postsRemoteRepository, postsLocalRepository);
        }
        return INSTANCE;
    }

    private PostsRepository(IPostsRepository postsRemoteRepository, IPostsRepository postsLocalRepository) {
        this.postsRemoteRepository = postsRemoteRepository;
        this.postsLocalRepository = postsLocalRepository;
    }

    @Override
    public void getPosts(final boolean forceUpdate, @NonNull final LoadPostsCallback callback) {
        if(forceUpdate){
            loadDataFromRemoteDataSource(forceUpdate, callback);
        } else {
            postsLocalRepository.getPosts(forceUpdate, new LoadPostsCallback() {
                @Override
                public void onPostsLoaded(List<Post> posts) {
                    if(posts.isEmpty()){
                        loadDataFromRemoteDataSource(forceUpdate, callback);
                    } else {
                        callback.onPostsLoaded(posts);
                    }
                }

                @Override
                public void onError(int statusCode) {
                    loadDataFromRemoteDataSource(forceUpdate, callback);
                }
            });
        }
    }

    @Override
    public void getPost(int postId, @NonNull LoadPostCallback callback) {
        postsLocalRepository.getPost(postId, callback);
    }

    private void loadDataFromRemoteDataSource(boolean forceUpdate, @NonNull final LoadPostsCallback callback) {
        postsRemoteRepository.getPosts(forceUpdate, new LoadPostsCallback() {
            @Override
            public void onPostsLoaded(List<Post> posts) {
                refreshLocalDataSource(posts);
                callback.onPostsLoaded(posts);
            }

            @Override
            public void onError(int statusCode) {
                callback.onError(statusCode);
            }
        });
    }

    private void refreshLocalDataSource(List<Post> posts) {
        postsLocalRepository.deleteAllData();
        postsLocalRepository.savePosts(posts);
    }

    @Override
    public void deleteAllData() {
        postsLocalRepository.deleteAllData();
    }

    @Override
    public void savePosts(List<Post> posts) {
        postsLocalRepository.savePosts(posts);
    }

    @VisibleForTesting
    static void destroyInstance() {
        INSTANCE = null;
    }
}

File: app/src/main/java/com/elvotra/clean/data/repository/PostsRepository.kt
package com.elvotra.clean.data.repository

import android.support.annotation.VisibleForTesting

import com.elvotra.clean.domain.model.Post
import com.elvotra.clean.domain.repository.IPostsRepository

import java.util.ArrayList

class PostsRepository private constructor(private val postsRemoteRepository: IPostsRepository, private val postsLocalRepository: IPostsRepository) : IPostsRepository {

    override fun getPosts(forceUpdate: Boolean, callback: IPostsRepository.LoadPostsCallback) {
        if (forceUpdate) {
            loadDataFromRemoteDataSource(forceUpdate, callback)
        } else {
            postsLocalRepository.getPosts(forceUpdate, object : IPostsRepository.LoadPostsCallback {
                override fun onPostsLoaded(posts: ArrayList<Post>) {
                    if (posts.isEmpty()) {
                        loadDataFromRemoteDataSource(forceUpdate, callback)
                    } else {
                        callback.onPostsLoaded(posts)
                    }
                }

                override fun onError(statusCode: Int) {
                    loadDataFromRemoteDataSource(forceUpdate, callback)
                }
            })
        }
    }

    override fun getPost(postId: Int, callback: IPostsRepository.LoadPostCallback) {
        postsLocalRepository.getPost(postId, callback)
    }

    private fun loadDataFromRemoteDataSource(forceUpdate: Boolean, callback: IPostsRepository.LoadPostsCallback) {
        postsRemoteRepository.getPosts(forceUpdate, object : IPostsRepository.LoadPostsCallback {
            override fun onPostsLoaded(posts: ArrayList<Post>) {
                refreshLocalDataSource(posts)
                callback.onPostsLoaded(posts)
            }

            override fun onError(statusCode: Int) {
                callback.onError(statusCode)
            }
        })
    }

    private fun refreshLocalDataSource(posts: ArrayList<Post>) {
        postsLocalRepository.deleteAllData()
        postsLocalRepository.savePosts(posts)
    }

    override fun deleteAllData() {
        postsLocalRepository.deleteAllData()
    }

    override fun savePosts(posts: ArrayList<Post>) {
        postsLocalRepository.savePosts(posts)
    }

    companion object {

        private var INSTANCE: PostsRepository? = null

        fun getInstance(postsRemoteRepository: IPostsRepository, postsLocalRepository: IPostsRepository): PostsRepository? {
            if (INSTANCE == null) {
                INSTANCE = PostsRepository(postsRemoteRepository, postsLocalRepository)
            }
            return INSTANCE
        }

        @VisibleForTesting
        internal fun destroyInstance() {
            INSTANCE = null
        }
    }
}
